horovod/common/ops/mpi_operations.cc

```diff
@@ -0,0 +1,405 @@
+    : AllreduceOp(global_state), mpi_context_(mpi_context) {}
+
+bool MPIAllreduce::Enabled(ParameterManager& param_manager,
+                           std::vector<TensorTableEntry>& entries,
+                           const Response& response) const {
+  return true;
+}
+
+void MPIAllreduce::DoAllreduce(std::vector<TensorTableEntry>& entries,
+                               const void* fused_input_data, void* buffer_data,
+                               int64_t& num_elements, size_t& buffer_len) {
+  global_state_->timeline.ActivityStartAll(entries, MPI_ALLREDUCE);
+  DoMPIAllreduce(mpi_context_, entries, buffer_data, num_elements, buffer_len);
+  global_state_->timeline.ActivityEndAll(entries);
+}
+
+#if HAVE_CUDA
+MPI_CUDAAllreduce::MPI_CUDAAllreduce(MPIContext* mpi_context,
+                                     HorovodGlobalState* global_state)
+                                     : CUDAAllreduce(cuda_context, comm_context, global_state),
+                                       mpi_context_(mpi_context) {}
+
+void MPI_CUDAAllreduce::DoAllreduce(std::vector<TensorTableEntry>& entries,
+                                    const void* fused_input_data, void* buffer_data,
+                                    int64_t& num_elements, size_t& buffer_len) {
+  global_state_->timeline.ActivityStartAll(entries, MPI_ALLREDUCE);
+  DoMPIAllreduce(mpi_context_, entries, buffer_data, num_elements, buffer_len);
+  global_state_->timeline.ActivityEndAll(entries);
+}
+#endif
+
+MPIAllgather::MPIAllgather(MPIContext* mpi_context, HorovodGlobalState* global_state)
+    : AllgatherOp(global_state), mpi_context_(mpi_context) {}
+
+bool MPIAllgather::Enabled(ParameterManager& param_manager,
+                           std::vector<TensorTableEntry>& entries,
+                           const Response& response) const {
+  return true;
+}
+
+void MPIAllgather::DoAllgather(std::vector<TensorTableEntry>& entries, int* recvcounts, int* displcmnts,
+                               int64_t** entry_component_offsets, int64_t** entry_component_sizes,
+                               int64_t total_size, int element_size) {
+  // Data is at the CPU and hierarchical allgather is disabled, or
+  // Data is at the GPU and HOROVOD_GPU_ALLGATHER == MPI
+  auto& timeline = global_state_->timeline;
+  auto& first_entry = entries[0];
+
+  const void* sendbuf = nullptr;
+  int64_t total_num_elements = 0;
+  void* buffer_data;
+
+  if (entries.size() > 1) {
+    auto& buffer = global_state_->fusion_buffer.GetBuffer(
+        first_entry.device, first_entry.context->framework());
+    buffer_data = const_cast<void*>(buffer->AccessData(first_entry.context));
+
+    // Copy memory into the fusion buffer. Then the input data of each
+    // process is assumed to be in the area where that process would
+    // receive its own contribution to the receive buffer.
+    timeline.ActivityStartAll(entries, MEMCPY_IN_FUSION_BUFFER);
```

Comment:
Should this logic be in abstract Allgather()?

Reply:
Good idea, done.